### PR TITLE
Treat warnings as errors

### PR DESCRIPTION
**Pull request by j178**
Fixes https://github.com/encode/httpcore/issues/384

Follow up #368

- Use `ssl.create_default_context()` instead `ssl.SSLContext()` to avoid some ssl deprecation warnings
- Add a `ssl_context` fixture that trusts our self issued certificates for localhost
- Minor fix for the enum repr change in 3.10
- Close the UDS socket if connection error

### CODE REVIEW

File: httpcore/_async/http2.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def __init__(
         self._exhausted_available_stream_ids = False
 
     def __repr__(self) -> str:
-        return f"<AsyncHTTP2Connection [{self._state}]>"
+        return f"<AsyncHTTP2Connection [{self._state.name}]>"
```
Comment (j178) [inline httpcore/_async/http2.py:54]:
```DeprecationWarning: in 3.12 format() will use the enum member, not the enum member's value;
use a format specifier, such as :d for an IntEnum member, to maintain the current display.
```

File: httpcore/_backends/sync.py [hunk 2]
```diff
@@ -155,13 +155,17 @@ def open_uds_stream(
...
+            try:
+                sock.settimeout(connect_timeout)
+                sock.connect(path)
+
+                if ssl_context is not None:
+                    sock = ssl_context.wrap_socket(
+                        sock, server_hostname=hostname.decode("ascii")
+                    )
+            except OSError:
+                sock.close()
```
Comment (j178) [inline httpcore/_backends/sync.py:167]:
```
ResourceWarning: unclosed <socket.socket fd=11, family=AF_UNIX, type=SOCK_STREAM, proto=0>
```

**Comment by graingert** [comment]
There's a PR here btw https://github.com/encode/httpcore/pull/364

File: httpcore/_async/connection.py [hunk 3]
```diff
@@ -33,7 +34,9 @@ def __init__(
         self._http2_enabled = http2
         self._keepalive_expiry = keepalive_expiry
         self._uds = uds
-        self._ssl_context = SSLContext() if ssl_context is None else ssl_context
+        self._ssl_context = (
+            ssl.create_default_context() if ssl_context is None else ssl_context
```
Comment (graingert) [inline httpcore/_async/connection.py:38]: create_default_context creates a secure context but SSLContext() creates an insecure one

Comment (graingert) [reply httpcore/_async/connection.py:38]: is it work moving https://github.com/encode/httpx/blob/b43af721cd7804214396b5511e4f68d144a9b6a6/httpx/_config.py#L44 into httpcore?

File: httpcore/_async/connection.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+import ssl
 from ssl import SSLContext
```
Comment (graingert) [inline httpcore/_async/connection.py:2]: would be nice to pick one import style and stick with it

File: httpcore/_backends/sync.py [hunk 5]
```diff
@@ -156,12 +156,16 @@ def open_uds_stream(
         with map_exceptions(exc_map):
             sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
             sock.settimeout(connect_timeout)
-            sock.connect(path)
```
Comment (graingert) [inline httpcore/_backends/sync.py:159]: you can still get a ResourceWarning if this socket is GCd after a KeyboardInterrupt

```
            try:
                sock.connect(path)
            except BaseException:
                sock.close()
                raise
```

File: tests/conftest.py [hunk 6]
```diff
@@ -140,6 +142,21 @@ def localhost_cert_private_key_file(
...
 
 
+@pytest.fixture(scope="session")
+def ssl_context(
+    cert_authority: trustme.CA, localhost_cert: trustme.LeafCert
+) -> ssl.SSLContext:
+    ssl_context = ssl.create_default_context()
+
+    if hypercorn is None:
+        ssl_context.load_verify_locations(certifi.where())
```
Comment (graingert) [inline tests/conftest.py:152]: create_default_context should already load the correct verify locations

File: tests/utils.py [hunk 7]
```diff
@@ -179,6 +179,7 @@ def http_proxy_server(proxy_host: str, proxy_port: int):
         finally:
             if proc is not None:
                 proc.kill()
+                proc.wait()
```
Comment (graingert) [inline tests/utils.py:182]:
```suggestion
                proc.communicate()
```

File: httpcore/_async/http2.py (hunk 1 above)
Comment (graingert) [reply httpcore/_async/http2.py:54]: https://www.python.org/dev/peps/pep-0663/ that's this right?

Comment (graingert) [reply httpcore/_async/http2.py:54]: I think we should change the AsyncHTTP2Connection to use the repr of its state

```suggestion
        return f"<AsyncHTTP2Connection [{self._state!r}]>"
```

Comment (j178) [reply httpcore/_async/http2.py:54]: `Enum.__repr__` has different behavior between python3.10+ and below, we'll get different output among python versions.

Comment (graingert) [reply httpcore/_async/http2.py:54]: That's working as designed imho

File: httpcore/_async/connection.py (hunk 3 above)
Comment (tomchristie) [reply httpcore/_async/connection.py:38]: I think we'd like to have a minimal-as-possible SSL context in `httpcore`, eg...

https://github.com/tomchristie/httpcore-the-directors-cut/blob/main/httpcore/_ssl.py


```python
import certifi
import ssl


def default_ssl_context() -> ssl.SSLContext:
    context = ssl.create_default_context()
    context.load_verify_locations(certifi.where())
    return context
```

Comment (graingert) [reply httpcore/_async/connection.py:38]: I don't think `certifi.where()` is needed anymore in py3.6+

File: tests/conftest.py (hunk 6 above)
Comment (tomchristie) [reply tests/conftest.py:152]: I don't know if we need `certifi` in *this particular case* or not, but just to help make this clearer for anyone following the conversation...

Python's `ssl.create_default_context()` doesn't hook into the system certificate store, so it's actually *not*  able to correctly validate https hostnames...

```python
>>> import ssl, socket
>>> context = ssl.create_default_context()
>>> sock = socket.create_connection(("www.example.com", 443))
>>> ssl_sock = context.wrap_socket(sock, server_hostname="www.example.com")
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 423, in wrap_socket
    session=session
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 870, in _create
    self.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)
```

Which is why certifi is needed (A third party package which bundles up Mozilla's certificate store of root certificates)...

```python
>>> import certifi, ssl, socket
>>> context = ssl.create_default_context()
>>> context.load_verify_locations(certifi.where())
>>> sock = socket.create_connection(("www.example.com", 443))
>>> ssl_sock = context.wrap_socket(sock, server_hostname="www.example.com")
```

**Comment by tomchristie** [comment]
Closing as per #364

**Pull request closed without merging**